Add missing transfers for sDAI deposits and withdraws

fhenneke opened the pull request:
Conversions DAI<>sDAI  are not correctly accounted for in the current slippage query. This has led to two settlements with computes slippage even though the slippage was actually small.

- https://etherscan.io/tx/0xdae82500c69c66db4e4a8c64e1d6a95f3cdc5cb81a5a00228ce6f247b9b8cefd
- https://etherscan.io/tx/0x8dc62a04eb6b42e571c19cccf7d6503c83f8b1add671ba5b5a9c4c8404b1042a

The reason is that `Withdraw` and `Deposit` events are only accompanied by _one_ transfer. But in terms  buffer changes they correspond to _two_ transfers.

[Tenderly](https://www.tdly.co/tx/mainnet/0xdae82500c69c66db4e4a8c64e1d6a95f3cdc5cb81a5a00228ce6f247b9b8cefd) and [eigenphi](https://eigenphi.io/mev/eigentx/0xdae82500c69c66db4e4a8c64e1d6a95f3cdc5cb81a5a00228ce6f247b9b8cefd) get this wrong as well in their asset change calculations. To see what happens to buffers one can use [phalcon](https://explorer.phalcon.xyz/tx/eth/0xdae82500c69c66db4e4a8c64e1d6a95f3cdc5cb81a5a00228ce6f247b9b8cefd) or look at state changes.

With the proposed change, additional transfers are created by looking at event logs of the SavingsDAI contract.
- One AMM_IN transfer is created for each `Withdraw` event.
- One AMM_OUT transfer is created for each `Deposit` event.

This is not a general solution and only works for sDAI. Since we want to support that token now, I added this fix to the query.
Long term it would be nice to have a more general solution.

fhenneke committed:
add missing transfers for cDAI deposits and withdraws

looks at event logs of the savings dai contract and adds a
- AMM_IN transfers for Withdraw events
- AMM_OUT transfers for Deposit events

bh2smith reviewed (approved):
Nice job here!

harisang commented on queries/dune_v2/period_slippage.sql:
```diff
@@ -125,6 +125,33 @@ batch_meta as (
     -- ETH transfers to traders are already part of USER_OUT
     and not contains(traders_out, to)
 )
+-- sDAI emit only one transfer event for deposits and withdrawls.
```
```suggestion
-- sDAI emits only one transfer event for deposits and withdrawals.
```

fhenneke committed:
added pull request number

fhenneke committed:
fixed typos

fhenneke commented:
I fixed some typos and added the pull request number.

For completeness, here is a forked query on dune for the two affected auctions:
- https://dune.com/queries/3286342?TxHash_t6c1ea=0xdae82500c69c66db4e4a8c64e1d6a95f3cdc5cb81a5a00228ce6f247b9b8cefd&EndTime_d83555=2023-12-07+00%3A00%3A00&StartTime_d83555=2023-12-06+00%3A00%3A00
- https://dune.com/queries/3286342?TxHash_t6c1ea=0x8dc62a04eb6b42e571c19cccf7d6503c83f8b1add671ba5b5a9c4c8404b1042a&EndTime_d83555=2023-12-09+00%3A00%3A00&StartTime_d83555=2023-12-07+00%3A00%3A00

If this PR is merged, I will update
- the original period slippage query
- the period payout query (for the dashboard)
- the unusual slippage query (for our dune alerts)

harisang commented:
Looks good! Doing some sanity checks with settlements with sDAI for which we didn't have any accounting issues, and if things look good, will approve

harisang commented:
Here is an example where we seem to have a problem.

https://dune.com/queries/3286342?EndTime_d83555=2023-11-27+00%3A00%3A00&StartTime_d83555=2023-11-24+00%3A00%3A00&TxHash_t6c1ea=0x3a86fbaf72ba3f15e8510ae2ac58e8fb16021544b492fdb4118fbaaadbd22aca

fhenneke committed:
only add transfers if settlement contract does the trading

fhenneke commented:
Good catch. The problem was that the tranfer ledger should only contain transfers to and from the settlement contract. Thus it also only should add additional transfers for sDAI if the settlement contract is doing the trading. I added a check in the query.

harisang commented:
All tests that I ran checked out, and the code makes sense, so I am approving!
